datafusion/core/src/physical_optimizer/pipeline_fixer.rs

```diff
@@ -182,13 +289,46 @@ fn apply_subrules_and_check_finiteness_requirements(
     physical_optimizer_subrules: &Vec<Box<PipelineFixerSubrule>>,
 ) -> Result<Option<PipelineStatePropagator>> {
     for sub_rule in physical_optimizer_subrules {
-        if let Some(value) = sub_rule(&input).transpose()? {
+        if let Some(value) = sub_rule(input.clone()).transpose()? {
             input = value;
         }
     }
     check_finiteness_requirements(input)
 }
```

Comment:
Regarding the method `apply_subrules_and_check_finiteness_requirements()`

I would suggest to make the `input` immutable. If there is changes to the plan/struct, return the changed (new)plan/struct.


```rust
fn apply_subrules_and_check_finiteness_requirements(
    mut input: PipelineStatePropagator,
    physical_optimizer_subrules: &Vec<Box<PipelineFixerSubrule>>,
)
```

Suggest change to

```rust
fn apply_subrules_and_check_finiteness_requirements(
    input: PipelineStatePropagator,
    physical_optimizer_subrules: &[Box<PipelineFixerSubrule>],
) -> Result<Option<PipelineStatePropagator>> {
    let after_op =
        physical_optimizer_subrules
            .iter()
            .try_fold(input, |pipeline, sub_rule| {
                if let Some(value) = sub_rule(&pipeline).transpose()? {
                    Result::<_, DataFusionError>::Ok(value)
                } else {
                    Result::<_, DataFusionError>::Ok(pipeline)
                }
            })?;
    check_finiteness_requirements(after_op)
}
```

Reply:
I tried it, but this creates an ownership problem in the closure (since `sub_rule` does not take a reference as its argument). We can make this work if we change the sub-rule return type so that it always returns an `PipelineStatePropagator` (by returning the argument unchanged when nothing is modified), and not an `Option`. Maybe we can explore this in a refactor PR. BTW, we came across a similar issue in some use cases involving `tranform_up`/`transform_down`. Maybe we can consider all of those in that PR.

Will keep this in mind, thanks for pointing it out.

Reply:
Sure, you can do all code refinement in following PR. 
BTW, in your use case, do you use Ballista as the distributed engine ?

Reply:
Good to know.  In the past, there was some discussion to enhance Ballista to support both BATCH/STREAMING execution models:

https://docs.google.com/document/d/1OdAe078axk4qO0ozUxNqBMD4wKoBhzh9keMuLp_jerE/edit#
http://www.vldb.org/pvldb/vol11/p746-yin.pdf
I haven't working on Ballista since last year and there is no progress in this area.

And in the latest Flink release, they had implement similar features(Bubble execution model, hybird shuffle etc).
https://cwiki.apache.org/confluence/display/FLINK/FLIP-235%3A+Hybrid+Shuffle+Mode

I think generally we can follow Flink's approach to make both DataFusion and Ballista support BATCH/STREAMING execution models.  In the high level, we can have different models(BATCH vs STREAMING), and user can specify the execution model. In the physical planing phase, we have `BatchPlanner` and `StreamingPlanner`, they can share some common rules, and batch and streaming planners can have their own rules. 
In the `ExecutionPlan` trait, we can have another trait to indicate some operators are `Source` operators, the source operators can be `BOUNDED` or `UNBOUNDED`.    `BOUNDED` or `UNBOUNDED` should be a property available to
Source operators only.

Reply:
We have made significant progress in foundational streaming execution support on the Datafusion side over the last few months. For example, you can already mark sources as infinite/unbounded (e.g. we have tests with FIFO files). Every `ExecutionPlan` then propagates this information upwards, so we make adjustments to the plan in an intelligent way depending on whether we are dealing with a stream.

We will write a blog post about this soon, I will share with you when it is done so you can get up to speed on what we've done already. I think it would be also good to have a meeting with everyone interested in this to discuss future plans and designs.